### PR TITLE
add znr_gl_base_technique_draw_arrays

### PR DESCRIPTION
**Pull request by watasuke102**
## Context

continued from #206 

## Summary

add znr_gl_base_technique_draw_arrays, send information about draw method

## How to check behavior

none

**This PR is based on #213**

### CODE REVIEW

File: zgnroot/include/zgnr/gl-base-technique.h [hunk 1]
```diff
@@ -8,6 +8,18 @@
 extern "C" {
 #endif
 
+enum zgnr_gl_base_technique_draw_method {
+  ZGNR_GL_BASE_TECHNIQUE_DRAW_ARRAYS,
+};
```
Comment (Aki-7) [inline zgnroot/include/zgnr/gl-base-technique.h:13]:
```suggestion
enum zgnr_gl_base_technique_draw_method {
  ZGNR_GL_BASE_TECHNIQUE_DRAW_METHOD_NONE,
  ZGNR_GL_BASE_TECHNIQUE_DRAW_METHOD_ARRAYS,
};
```

We need an initial state of `zgnr_gl_base_technique::current.draw_method`, which means no drawing method specified yet.

File: zgnroot/src/gl-base-technique.c [hunk 2]
```diff
@@ -118,10 +119,12 @@ zgnr_gl_base_technique_protocol_draw_arrays(struct wl_client *client,
     struct wl_resource *resource, uint32_t mode, int32_t first, uint32_t count)
 {
   UNUSED(client);
-  UNUSED(resource);
-  UNUSED(mode);
-  UNUSED(first);
-  UNUSED(count);
+  struct zgnr_gl_base_technique_impl *self =
+      wl_resource_get_user_data(resource);
+  self->pending.draw_method = ZGNR_GL_BASE_TECHNIQUE_DRAW_ARRAYS;
```
Comment (Aki-7) [inline zgnroot/src/gl-base-technique.c:124]:
```suggestion
  struct zgnr_gl_base_technique_impl *self =
      wl_resource_get_user_data(resource);
  if (self == NULL) return;
      
  self->pending.draw_method = ZGNR_GL_BASE_TECHNIQUE_DRAW_ARRAYS;
```

Regarding base technique, it's wl_resource can be inert.
Inert resource means, the resoruce object itself is alive but there is no bound object. That is, wl_resource_get_user_data(resource) is NULL.

As for base technique, that happens when rendering unit bound to the base technique is destroyed.

If you need more explanation, feel free to ask me at any time!

File: zgnroot/src/gl-base-technique.c [hunk 3]
```diff
@@ -178,12 +181,23 @@ zgnr_gl_base_technique_handle_rendering_unit_commit(
...
-  if (vertex_array) {
-    zgnr_gl_vertex_array_commit(vertex_array);
-    zgnr_gl_base_technique_set_current_vertex_array(self, &vertex_array->base);
-  } else {
-    zgnr_gl_base_technique_set_current_vertex_array(self, NULL);
+  self->base.current.args = self->pending.args;
+  self->base.current.draw_method = self->pending.draw_method;
+
+  self->base.current.args = self->pending.args;
+  self->base.current.draw_method = self->pending.draw_method;
```
Comment (Aki-7) [inline zgnroot/src/gl-base-technique.c:188]: DRY :)

File: zna/scene/virtual-object/gl-base-technique.c [hunk 4]
```diff
@@ -58,9 +58,22 @@ zna_gl_base_technique_apply_commit(
...
 
-  // TODO: Apply draw arrays, etc.
+  switch (self->zgnr_gl_base_technique->current.draw_method) {
+    case ZGNR_GL_BASE_TECHNIQUE_DRAW_ARRAYS:
+      znr_gl_base_technique_draw_arrays(self->znr_gl_base_technique,
+          self->zgnr_gl_base_technique->current.args.arrays.mode,
+          self->zgnr_gl_base_technique->current.args.arrays.first,
+          self->zgnr_gl_base_technique->current.args.arrays.count);
+      break;
+  }
```
Comment (Aki-7) [inline zna/scene/virtual-object/gl-base-technique.c:76]: Summing up above discussion, we need to

1. check that the current draw method has been changed or not.
2. be careful of the initial state

You also have to consider about the situation where only_damaged == false, which means session has been reconnected and we need to do `znr_gl_base_technique_draw_ararys` or similar even if the draw method has not been modified.

File: zgnroot/include/zgnr/gl-base-technique.h [hunk 5]
```diff
@@ -8,6 +8,19 @@
 extern "C" {
 #endif
 
+enum zgnr_gl_base_technique_draw_method {
+  ZGNR_GL_BASE_TECHNIQUE_DRAW_NONE,
+  ZGNR_GL_BASE_TECHNIQUE_DRAW_ARRAYS,
```
Comment (Aki-7) [inline zgnroot/include/zgnr/gl-base-technique.h:13]:
```suggestion
  ZGNR_GL_BASE_TECHNIQUE_DRAW_METHOD_NONE,
  ZGNR_GL_BASE_TECHNIQUE_DRAW_METHOD_ARRAYS,
```

File: zgnroot/include/zgnr/gl-base-technique.h [hunk 6]
```diff
@@ -20,6 +33,10 @@ struct zgnr_gl_base_technique {
   struct {
     struct zgnr_gl_vertex_array *vertex_array;  // nullable
     bool vertex_array_changed;
+
+    enum zgnr_gl_base_technique_draw_method draw_method;
+    union zgnr_gl_base_technique_draw_args args;
+    bool draw_info_changed;
```
Comment (Aki-7) [inline zgnroot/include/zgnr/gl-base-technique.h:39]:
```suggestion
    bool draw_method_changed;
```

"draw info" seems a bit too broad in meaning. This could include vertex array, texture, etc.

File: zgnroot/src/gl-base-technique.c [hunk 7]
```diff
@@ -238,11 +250,14 @@ zgnr_gl_base_technique_create(struct wl_client *client, uint32_t id,
...
   self->base.comitted = false;
+  self->base.current.draw_method = ZGNR_GL_BASE_TECHNIQUE_DRAW_NONE;
+  self->base.current.draw_info_changed = false;
+  self->pending.draw_info_changed = false;
+
   self->base.current.vertex_array = NULL;
   self->base.current.vertex_array_changed = false;
 
   wl_signal_init(&self->base.events.destroy);
-
```
Comment (Aki-7) [inline zgnroot/src/gl-base-technique.c:245]: Why did you remove this line? It's just a mistake or any intension?

File: zgnroot/include/zgnr/gl-base-technique.h (hunk 6 above)
Comment (watasuke102) [reply zgnroot/include/zgnr/gl-base-technique.h:39]: > "draw info" seems a bit too broad in meaning

Yes, I agree this. but I think when call `draw_array` twice, but draw method isn't changed.
So I named this `draw_info` instead of `draw_method` to inform that this monitors both method and args.

File: zgnroot/src/gl-base-technique.c (hunk 7 above)
Comment (watasuke102) [reply zgnroot/src/gl-base-technique.c]: This is my mistake; occured when rebasing to main, I guess

**Review by Aki-7 — APPROVED** [review]
